src/tests/ftest/telemetry/dkey_akey_enum_punch.py

```diff
@@ -0,0 +1,371 @@
+        straightforward.
+
+        Latency is the time it took to process the list_dkey() calls. The slight
+        difference from active is that min would contain the actual latency and
+        not 0 when there's at least one call.
+
+        6. akey enum active and latency
+        engine_io_ops_akey_enum_active_max
+        engine_io_ops_akey_enum_active_mean
+        engine_io_ops_akey_enum_active_min
+        engine_io_ops_akey_enum_active_stddev
+        engine_io_ops_akey_enum_latency_max
+        engine_io_ops_akey_enum_latency_mean
+        engine_io_ops_akey_enum_latency_min
+        engine_io_ops_akey_enum_latency_stddev
+
+        Same idea as dkey. We also expect to see very similar output as in dkey.
+
+        7. dkey punch active and latency
+        engine_io_ops_dkey_punch_active_max
+        engine_io_ops_dkey_punch_active_mean
+        engine_io_ops_dkey_punch_active_min
+        engine_io_ops_dkey_punch_active_stddev
+        engine_io_ops_dkey_punch_latency_max
+        engine_io_ops_dkey_punch_latency_mean
+        engine_io_ops_dkey_punch_latency_min
+        engine_io_ops_dkey_punch_latency_stddev
+
+        Same as 5 except we're measuring the active and latency for
+        punch_dkeys() calls.
+
+        8. akey punch active and latency
+        engine_io_ops_akey_punch_active_max
+        engine_io_ops_akey_punch_active_mean
+        engine_io_ops_akey_punch_active_min
+        engine_io_ops_akey_punch_active_stddev
+        engine_io_ops_akey_punch_latency_max
+        engine_io_ops_akey_punch_latency_mean
+        engine_io_ops_akey_punch_latency_min
+        engine_io_ops_akey_punch_latency_stddev
+
+        Same idea as dkey. We also expect to see very similar output as in dkey.
+
+        :avocado: tags=all,full_regression
+        :avocado: tags=vm
+        :avocado: tags=telemetry
+        :avocado: tags=dkey_akey_enum_punch
+        """
+        self.set_num_targets()
+
+        self.add_pool()
+        container = DaosContainer(self.context)
+        container.create(self.pool.pool.handle)
+        container.open()
+
+        # Object type needs to be OC_S1 so that the objects are spread across
+        # all targets.
+        self.write_objects_insert_keys(
+            container=container, objtype=DaosObjClass.OC_S1)
+
+        # Call list_dkey() and list_akey() on each object.
+        for i in range(self.obj_count):
+            _ = self.ioreqs[i].list_dkey()
+            _ = self.ioreqs[i].list_akey(dkey=self.dkeys_a[i])
+            _ = self.ioreqs[i].list_akey(dkey=self.dkeys_b[i])
+
+        self.punch_all_keys()
+
+        self.telemetry.dmg.verbose = False
+
+        ### Obtain and verify the io metrics 1 to 4. ###
+        pool_dkey_enum = self.telemetry.ENGINE_POOL_METRICS[5]
+        pool_akey_enum = self.telemetry.ENGINE_POOL_METRICS[2]
+        pool_dkey_punch = self.telemetry.ENGINE_POOL_METRICS[6]
+        pool_akey_punch = self.telemetry.ENGINE_POOL_METRICS[3]
```

Comment:
How do we know what indices 5,2,6,3 correspond to without looking at the implementation?

Reply:
Not sure what you mean. If it's the readability issue, I added the whole metric name above each line as a comment.

Reply:
It's more of a self-documentation issue. Not with this usage, but the implementation/definition of ENGINE_POOL_METRICS. It seems strange that ENGINE_POOL_METRICS is a list where each index contains a different metric. I would think ENGINE_POOL_METRICS should be a struct/class.
For example, taken out of context, ENGINE_POOL_METRICS[3] gives no indication about the metric it corresponds to. "3" is arbitrary The comments are helpful HERE, but any other usage of ENGINE_POOL_METRICS would run into a similar problem.
Anyway, your added comments answered my question - the actual implementation/definition of ENGINE_POOL_METRICS should probably be done separately (if we decide to change it)

Reply:
I see. I think that your struct/class approach would be more helpful than the current simple lists approach because ~70-80% of the metrics have min/max/mean/stddev, they're mostly divided into active and latency, and so on. There are clearly some patterns. Although I guess we should be careful not to over-engineer :)